packages/ui/src/components/Visualization/Custom/Node/CustomNode.tsx

```diff
@@ -27,146 +39,205 @@ import { NodeContextMenuFn } from '../ContextMenu/NodeContextMenu';
-            width={CanvasDefaults.ADD_STEP_ICON_SIZE}
-            height={CanvasDefaults.ADD_STEP_ICON_SIZE}
+            x={labelX}
+            y={boxRef.current.height - 1}
+            width={CanvasDefaults.DEFAULT_LABEL_WIDTH}
+            height={CanvasDefaults.DEFAULT_LABEL_HEIGHT}
+            className="custom-node__label"
           >
-            <AddStepIcon
-              vizNode={vizNode}
-              mode={AddStepMode.AppendStep}
-              title="Add step"
-              data-testid="quick-append-step"
+            <div
+              className={clsx('custom-node__label__text', {
+                'custom-node__label__text__error': doesHaveWarnings,
+              })}
             >
-              <Icon size="lg">{isHorizontal ? <ArrowRightIcon /> : <ArrowDownIcon />}</Icon>
-            </AddStepIcon>
+              {doesHaveWarnings && (
+                <Icon title={validationText} data-warning={doesHaveWarnings}>
+                  <ExclamationCircleIcon />
+                </Icon>
+              )}
+              <span title={label}>{label}</span>
+            </div>
           </foreignObject>
-        )}
 
-        {childCount && <LabelBadge badge={`${childCount}`} x={0} y={0} />}
-      </g>
-    </Layer>
-  );
-});
+          {!droppable && shouldShowToolbar && (
+            <Layer id={TOP_LAYER}>
+              <foreignObject
+                ref={toolbarHoverRef}
+                className="custom-node__toolbar"
+                x={toolbarX}
+                y={toolbarY}
+                width={toolbarWidth}
+                height={CanvasDefaults.STEP_TOOLBAR_HEIGHT}
+              >
+                <StepToolbar
+                  data-testid="step-toolbar"
+                  vizNode={vizNode}
+                  isCollapsed={element.isCollapsed()}
+                  onCollapseToggle={onCollapseToggle}
+                />
+              </foreignObject>
+            </Layer>
+          )}
+
+          {!droppable && shouldShowAddStep && (
+            <foreignObject
+              x={boxRef.current.width - 8}
+              y={(boxRef.current.height - CanvasDefaults.ADD_STEP_ICON_SIZE) / 2}
+              width={CanvasDefaults.ADD_STEP_ICON_SIZE}
+              height={CanvasDefaults.ADD_STEP_ICON_SIZE}
+            >
+              <AddStepIcon
+                vizNode={vizNode}
+                mode={AddStepMode.AppendStep}
+                title="Add step"
+                data-testid="quick-append-step"
+              >
+                <Icon size="lg">{isHorizontal ? <ArrowRightIcon /> : <ArrowDownIcon />}</Icon>
+              </AddStepIcon>
+            </foreignObject>
+          )}
+
+          {childCount && <LabelBadge badge={`${childCount}`} x={0} y={0} />}
+        </g>
+      </Layer>
+    );
+  },
+);
 
-export const CustomNodeWithSelection = withContextMenu(NodeContextMenuFn)(CustomNode);
+export const CustomNodeWithSelection = withDndDrop(customNodeDropTargetSpec)(
```

Comment:
For another PR: we can rename this component to CustomNode
```suggestion
export const CustomNode = withDndDrop(customNodeDropTargetSpec)(
```